### ISSUE TITLE
Building composite project with resolveJsonModule fails

### ISSUE THREAD

**Issue by strax** [Microsoft/TypeScript#25636]
**TypeScript Version:**  3.0.0-rc

**Search Terms:** 
resolveJsonModule, composite project, TS6307

**Code**

Repro at https://github.com/strax/ts-3-json-bug-repro

**Expected behavior:**

With `composite: true` and `resolveJsonModule: true` set in `tsconfig.json`, JSON modules are resolved in the compilation like they are when the `composite` compiler option is not set.

**Actual behavior:**

Running `tsc` gives the following error:

```
error TS6307: File '/src/hello.json' is not in project file list. Projects must list all files or use an 'include' pattern.
```

In addition, trying to include `src/hello.json` in the `files` property fails with

```
error TS6054: File '/src/hello.json' has unsupported extension. The only supported extensions are '.ts', '.tsx', '.d.ts'.
```

**Playground Link:**  

**Related Issues:**

**Comment by mhegazy** [comment]
we should allow `.json` in files list if `--resolveJsonModules` is enabled.

**Comment by sheetalkamat** [comment]
@mhegazy Is assumption that we shouldnt include them through include pattern of config file though?

**Comment by mhegazy** [comment]
yeah. i think that would be a bigger breaking change.

**Comment by cockscomb** [comment]
@strax @sheetalkamat @mhegazy Is this problem fixed in TypeScript 3.1.1?

I have tried to use TypeScript 3.1.1 in the [repro](https://github.com/strax/ts-3-json-bug-repro) project, TypeScript gives the following error:

```
error TS6307: File '/src/hello.json' is not in project file list. Projects must list all files or use an 'include' pattern.
```

It seems to this problem is still exist.

**Comment by sheetalkamat** [comment]
@cockscomb you need to have files clause to have the json file(`"files": ["src/hello.json"]` in the tsconfig file).  The issue here was it reported unsupported extension (which is not what you are seeing)
In your
cc: @RyanCavanaugh

**Comment by raymondfeng** [comment]
@sheetalkamat We ran into the similar issue as @cockscomb. `tsc` complains about json files not in project file list no matter how we tweak `tsconfig.json`:

```json
{
  "$schema": "http://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "rootDir": "src",
    "composite": true,
    "target": "es2017",
    "outDir": "dist"
  },
  "references": [
    ...
  ],
  "include": [
    "**/*"
  ]
}
```

None of the following works:
- "include": ["**/*"]
- "include": ["**/*.json", "**/*.ts"] 

Do we have to use `files` to include json files?

**Comment by sheetalkamat** [comment]
We do not want to add json files in compilation accidently and want users to explicitly specify them instead of using the include/exclude pattern.

**Comment by bajtos** [comment]
@sheetalkamat Thank you for the comment, I see your point.

Let me explain my use case. In LoopBack, we have a concept of a data source - a class providing access to a database, web service, etc. An application (a LoopBack/TypeScript project) can have multiple datasources. Each datasource is implemented in two files: a JSON file with the configuration (to make it easy to edit programmatically) and a TypeScript file exporting the datasource class.

The TypeScript code in `src/datasources/db.datasource.ts` ([full version](https://github.com/strongloop/loopback-next/blob/f9c9c6d6d5eb2b77df641e634a41e86c4b345909/examples/todo/src/datasources/db.datasource.ts
)):

```ts
import * as config from './db.datasource.json';

export class DbDataSource extends juggler.DataSource {
  constructor() {
    super(dsConfig);
  }
}
```

JSON config in `src/datasources/db.datasource.json` ([full version](https://github.com/strongloop/loopback-next/blob/f9c9c6d6d5eb2b77df641e634a41e86c4b345909/examples/todo/src/datasources/db.datasource.json)):

```ts
{
  "name": "db",
  "connector": "memory"
}
```

Relevant parts of tsconfig (we are not using project references yet):

```
{
  "$schema": "http://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "rootDir": "."
    "resolveJsonModule": true,

    "lib": ["es2018", "dom", "esnext.asynciterable"],
    "module": "commonjs",
    "moduleResolution": "node"
  },
  "include": ["index.ts", "src", "test"]
}
```

Now back to the topic of this issue:

1. When the project is _not_ composite and compiler is running in the "classic" mode (without `-b`), it correctly understands that `db.datasource.ts` is importing `db.datasource.json` and therefore the JSON file needs to be included in the compilation too.

    I find it confusing that composite/build mode changes this behavior.

2. When the project is switched to composite mode (the only change being adding `composite: true` to `compilerConfig`), the compilation starts failing with error `TS6307`. Despite this error, the actual JSON file is correctly copied to the output directory.

    I find it confusing that the compiler understands that a JSON file is imported and thus needs to be copied to the output directory, yet it complains that it was not explicitly listed in tsconfig's `files` section.

3. I find the requirement to explicitly add all JSON files one by one as a poor user experience. For applications (projects) containing many JSON files, the `tsconfig` content can become large and difficult to maintain.

If my first two arguments above are not convincing enough and/or there are other reasons explaining why JSON files must be treated differently in composite vs. non-composite projects: Can we at least allow JSON files to be specified via an explicit pattern in `include`, e.g. `src/datasources/*.datasource.json`? By explicit I mean that the pattern must match JSON files only (i.e. the pattern string must end with `.json`).

|pattern|file|match?|
|:--:|:--:|:--:|
`**/*` | `src/foo.ts` | ✔️  
`**/*` | `src/data.json` | ✖️ 
`src/*.json` | `src/foo.ts` | ✖️ 
`src/*.json` | `src/data.json` | ✔️

**Comment by sheetalkamat** [comment]
> I find it confusing that the compiler understands that a JSON file is imported and thus needs to be copied to the output directory, yet it complains that it was not explicitly listed in tsconfig's files section.

The issue isn't about including json in compiling but its about not specifying the json file in the list of inputs. Composite projects need explicitly include all files in the compilation as referencing projects need to be able to identify its all output.

Adding @DanielRosenwasser for opinion. The requirement you are suggesting is quite a work and need to be handled by multiple hosts as well (since it needs to integrate with our readDirectory interface as well). Apart from that right now if you have js and ts file with same name in location, the jsFile is ignored(since its lower in priority than ts file and that means that behavior needs change for json modules as well)

**Comment by sheetalkamat** [comment]
Ok I spoke too soon. I think we are able to achieve what you want and I will put up PR for this.

**Comment by bajtos** [comment]
Thank you @sheetalkamat for finding a solution and landing it via #27844 🙇 

I think it would be great to describe the interaction between `composite: true` and `resolveJsonModule` in the handbook too, e.g. https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Project%20References.md, and explain users how to configure their project to avoid "Projects must list all files or use an 'include' pattern." error. I will try to find some time to contribute those changes unless somebody else can contribute them quicker.

**Comment by EnixCoda** [comment]
This works for me:

```json
{
  "include": ["./src/**/*", "./src/**/*.json"],
}
```

**Comment by DylanVann** [comment]
There is still the issue of JSON files in `node_modules` not working.

**Comment by StrongerSY** [comment]
Listing JSON files in files section of tsconfig works for me ~ Thx @bajtos

**Comment by xtianus79** [comment]
This is still an issue

**Comment by bsunderhus** [comment]
I'm experiencing the same issue my project is setup as followed:

```
.
├── @types
├── package.json
├── src
│   ├── consts.json
│   ├── main
│   │   ├── index.ts
│   │   └── tsconfig.json
│   ├── tsconfig.json
│   ├── utils
│   │   ├── html-template.ts
│   │   └── tsconfig.json
│   └── worker
│       ├── state.ts
│       └── tsconfig.json
└── tsconfig.base.json
```

My main entry configuration file is `src/tsconfig.json` which only have references to other sub projects localized in `src/main`, `src/worker` and `src/utils` and in the `main` project I'm trying to access the `src/consts.json` file and I'm getting the error:

```log
File '/workspace/src/consts.json' is not listed within the file list of project '/workspace/src/main/tsconfig.json'. Projects must list all files or use an 'include' pattern.
``` 
Although I've listed it explicitly through `files` in `src/main/tsconfig`

```json
{
  "extends": "../../tsconfig.base.json",
  "files": [
    "./index.ts",
    "../consts.json",
  ],
  "references": [
    {"path": "../worker"},
    {"path": "../utils"}
  ]
}
```

**Comment by datrinh** [comment]
Still have this issue...

**Comment by trusktr** [comment]
I was spending time fiddling around with this until I realized it was a TS issue.

It is confusing and not intuitive to have to write something like

```json
{
  "include": ["./src/**/*", "./src/**/*.json"],
}
```

**Comment by PaulRBerg** [comment]
@trusktr's solution worked, but I had to restart my IDE.

Why is this issue closed?

**Comment by amerllica** [comment]
Just add this to the root of `tsconfig.json`:

```
{
  "include": ["**/*.json"],
}
```

**Comment by zerubeus** [comment]
this still an issue 2021 ... why this been closed?

**Comment by jpgupta** [comment]
@zerubeus - I just checked after running into this issue now, it seem like @TsaiTsaiChieh 's PR above fixes the issue by allowing a glob pattern in tsconfig's include property:

include: ["**/*.json"] 

https://github.com/TsaiTsaiChieh/never-stray/commit/ac22f38c00a0a7f997ee3ca1a7c9eb6b90feee50

Confirmed working on typescript 4.5.4

**Comment by SalahAdDin** [comment]
No any of the given solutions works for us.

**Comment by bingtimren** [comment]
This issue should be re-opened. The "closed" status probably will direct people away from this post, and make them waste many more hours where the solution is just in here.

For me, I have spent a few hours, and nearly tear down all the tsconfig settings in a mono-repo to try to solve this. I wish I had come to this issue and saw @trusktr 's comment earlier.

So just re-confirm,  
- `"include": [    "src/**/*" ]`  does NOT include json, 
- must use  `"include": [    "src/**/*", "src/**/*.json"]`

Maybe the reason this issue was closed, is because the initial issue report, while described the sympton correctly, had a misleading interpretation

> Expected behavior:
> 
> With composite: true and resolveJsonModule: true set in tsconfig.json, JSON modules are resolved in the compilation like they are when the composite compiler option is not set.
> 

Actually, composite and resolveJsonModule does not change the behaviour whether or not JSON is included in the INITIAL set of files - this can be confirmed by invoking tsc with --showConfig option

With or without composite and resolveJsonModule settings, the JSON files are not in the calculated "files" list.

However, when composite is false, even the JSON files are not in the initial file list, tsc does not care, as long as the files can be resolved eventually. 

When composite is true, to achieve composite build, tsc requires all imported file must be included in the INITIAL set of files. Then the issue becomes an error.

The the root cause, "*/**" does not include JSON, is always there, not solved. Only when composite is true, it suffaces and become a problem. Otherwise it's under the water.
